include/circt/Support/PrettyPrinter.h

```diff
@@ -0,0 +1,298 @@
+//===- PrettyPrinter.h - Pretty printing ----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This implements a pretty-printer.
+// "PrettyPrinting", Derek C. Oppen, 1980.
+// https://dx.doi.org/10.1145/357114.357115
+//
+// This was selected as it is linear in number of tokens O(n) and requires
+// memory O(linewidth).
+//
+// See PrettyPrinter.cpp for more information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_PRETTYPRINTER_H
+#define CIRCT_SUPPORT_PRETTYPRINTER_H
+
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/SaveAndRestore.h"
+
+#include <cstdint>
+#include <deque>
+#include <limits>
+
+namespace circt {
+namespace pretty {
+
+//===----------------------------------------------------------------------===//
+// Tokens
+//===----------------------------------------------------------------------===//
+
+/// Style of breaking within a group:
+/// - Consistent: all fits or all breaks.
+/// - Inconsistent: best fit, break where needed.
+/// - Never: force no breaking including nested groups.
+enum class Breaks { Consistent, Inconsistent, Never };
+
+/// Style of indent when starting a group:
+/// - Visual: offset is relative to current column.
+/// - Block: offset is relative to current base indentation.
+enum class IndentStyle { Visual, Block };
+
+class Token {
+public:
+  enum class Kind { String, Break, Begin, End };
+  struct StringInfo {
+    const char *str;
+    uint32_t len;
```

Comment:
There are some casts of `size_t` to `uint32_t` later in the code. It is possible that some of this would simplify by using `size_t` in place of `uint32_t` given that LLVM is using this as the return value/storage for a `StringRef`'s size. Only for your consideration.